### ISSUE TITLE
Kubelet regularly freeze control groups causing issues further down

### ISSUE THREAD

**Issue by odinuge** [kubernetes/kubernetes#104280]
#### What happened:

On cgroup v1 when using systemd cgroup driver, kubelet regularly freeze and thaw its control groups, and that is causing issues further down.

#### What you expected to happen:

No freeze

#### How to reproduce it (as minimally and precisely as possible):

Running the command below should never print anything, since freezer.state should always be `THAWED`.

```
# zsh globs
while true ; do cat /sys/fs/cgroup/freezer/kubepods.slice/{*/{*/,},}freezer.state; done | grep FROZEN
```

However, when using master/v1.22.0, this returns `FROZEN` regularly.

When using runc v1.0.0 or earlier, this can cause containers to end in a permanently frozen state. This state is tricky to debug, since no log output is given. To see if an environment is having problems, one can use:

```
# zsh syntax
$ grep -R "FROZEN"  /sys/fs/cgroup/freezer/kubepods.slice/**/freezer.state
# If there is no output, there is no issue. If there is output one or more container are frozen.
```

#### Anything else we need to know?:

More info in https://github.com/kubernetes/kubernetes/issues/102676 and https://github.com/kubernetes/kubernetes/issues/102676.

#### Environment:
- Kubernetes version (use `kubectl version`):
  - `v1.22.0`: introduced in https://github.com/kubernetes/kubernetes/pull/103743
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

**Comment by liggitt** [comment]
Thanks for opening this. IIUC, this issue is limited to:
* cgroup v1
* systemd cgroup driver
* runc 1.0.1+

**Comment by odinuge** [comment]
> Thanks for opening this. IIUC, this issue is limited to:
>
>- cgroup v1
>- systemd cgroup driver
>- runc 1.0.1+

Well, yeah, kinda. It is a bit of a matrix

As long as

- cgroup version = v1
- cgroup driver = systemd

the freezing will occur. The freezing is unnecessary, and creates a big overhead on machines with huge number of processes.

However, if runc (the binary used by the container runtime) is v1.0.0 or before, the permanent freeze of containers can occur. If using runc (the binary) v1.0.1, the permanent freeze is mitigated and does not occur. So in one way, as long as the user is using runc binary v1.0.1 or later, this is not completely critical.

**Comment by harche** [comment]
@odinuge 

I am seeing a lot of flakiness lately in systemd node e2e jobs using crio, 
https://testgrid.k8s.io/sig-node-cri-o#ci-crio-cgroupv1-node-e2e-conformance

Almost all of them fail with container not being able to come up with following error, 
`cannot start a container in the paused state` 

That seems to be coming from https://github.com/opencontainers/runc/blob/0d49470392206f40eaab3b2190a57fe7bb3df458/start.go#L51

I am wondering if that flakiness is due to frozen cgroups issue reported in this issue. Unfortunately I do no have access to the system to query sysfs to see if cgroups is in FROZEN state.

**Comment by liggitt** [comment]
cc @mrunalp @derekwaynecarr

**Comment by odinuge** [comment]
Yes @harche, that looks like a result of this issue. Thanks for the detailed information!


@saschagrunert:

> The code in runc seems intentional and exists since a long time, are we aware of any recent changes in k/k to this kind of cgroup handling?

This was issue in k8s (the freezing) was introduced in https://github.com/kubernetes/kubernetes/pull/103743. As disucssed here: https://github.com/kubernetes/kubernetes/issues/102676#issuecomment-880653430, freezing would cause issues, and the runc bump was merged before it was properly tested. If you are talking about runc as in the binary, there was previously an issue with the definition of a "paused/frozen" container; https://github.com/opencontainers/runc/pull/3065 should give some more context. That issue is fixed in runc v1.0.1 (the binary).

**Comment by odinuge** [comment]
This issue will be fixed by https://github.com/opencontainers/runc/pull/3143 or any similar way to avoid the freeze.

**Comment by mrunalp** [comment]
cc: @kolyshkin
